[quantization] Memoize parseIR calls in graph mode quantization

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #30199 [quantization] Use ConvPackedParams everywhere
* #30198 [quantization] Use LinearPackedParams everywhere
* **#30188 [quantization] Memoize parseIR calls in graph mode quantization**
* #30264 [quantization] Dont use SubgraphRewriter in FoldQuantizeCallIntoBuffer

We don't need to re-parse the IR on every call for these sites.

followup: can we extend this to places we use SubgraphMatcher?

Differential Revision: [D18625743](https://our.internmc.facebook.com/intern/diff/D18625743)

Commit:
[quantization] Memoize parseIR calls in graph mode quantization

[ghstack-poisoned]

Commit:
Update on "[quantization] Memoize parseIR calls in graph mode quantization"


We don't need to re-parse the IR on every call for these sites.

followup: can we extend this to places we use SubgraphMatcher?

Differential Revision: [D18625743](https://our.internmc.facebook.com/intern/diff/D18625743)

[ghstack-poisoned]

Commit:
Update on "[quantization] Memoize parseIR calls in graph mode quantization"


We don't need to re-parse the IR on every call for these sites.

followup: can we extend this to places we use SubgraphMatcher?

Differential Revision: [D18625743](https://our.internmc.facebook.com/intern/diff/D18625743)

[ghstack-poisoned]

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -19,10 +19,24 @@ namespace torch {
 namespace jit {
 namespace {
 
+struct PatternInfo {
```
Please add some comments describing this class as well, Thanks!

jerryzh168 reviewed (approved):
LGTM

jerryzh168 commented:
cc @ZolotukhinM do we want to move this change to SubgraphMatcher?

ZolotukhinM commented:
> cc @ZolotukhinM do we want to move this change to SubgraphMatcher?

SubgraphMatcher doesn't do any parsing, it takes graphs, so having this logic where it is now seems fine to me. It's up to the user of the matcher to figure out how to store the patterns and how to avoid reparsing the same thing several times.

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -262,28 +275,29 @@ void InsertObserversHelper::addIntermediateValuesToSkipObserver(
   // Note that the name of the value we want to skip inserting observer for
   // is hard coded as "intermediate_val"
-  std::string conv_functional_relu = R"(
+  static const PatternInfo conv_functional_relu =
```
I'm not sure I like making them static - the parsed graphs will sit in memory forever.

jamesr66a replied:
Yeah that's definitely true. My thought is that a fixed space cost is better than having a runtime cost that scales with the size of the model (e.g. some auto-generated code has a huge number of models and the users want to quantize it). How about we just stay on the lookout for if people complain about too much space being taken up?

ZolotukhinM replied:
Could we make it local to the `InsertObserversHelper`? This class handles the entire module, so in terms of runtime costs we would only parse everything once per a model. At the same time, once we finish quantizing the model, we would free all of it and not take any memory. The problem with staying on the lookout is that it is a very small increase - one of those that tend to creep in over time and are almost impossible to detect, but accumulative effect might be significant.

ZolotukhinM replied:
Another point is that the runtime costs are only for the "quantize" call - that's supposedly only run once and ahead-of-time, when performance is not critical. Memory usage, on contrary, would affect everything, including actual runs of models.

jamesr66a replied:
Here's my assesment:

1) callsites in InsertObserversHelper - yeah we can definitely make these class-local
2) FoldConvBatchNorm2d - Free function, maybe we can just remove the caching--cost doesn't scale
3) FoldQuantizeCallIntoBuffer - Free function, maybe we can just remove the caching--cost doesn't scale
4) FoldPrepackedWeightIntoModule - free function called for each method. We could probably wrap the two overloads of this function into a struct and save the compiled patterns there

jamesr66a commented:
@ZolotukhinM the subgraph rewriter does parsing here: https://github.com/pytorch/pytorch/blob/1f5105128780e5e32d6e9eb3f6787c9330400890/torch/csrc/jit/passes/subgraph_rewrite.cpp#L59

ZolotukhinM commented:
> @ZolotukhinM the subgraph rewriter does parsing here:
> 
> https://github.com/pytorch/pytorch/blob/1f5105128780e5e32d6e9eb3f6787c9330400890/torch/csrc/jit/passes/subgraph_rewrite.cpp#L59

The places you changed used SubgraphMatcher directly, without SubgraphRewriter, If I didn't miss something.

SubgraphRewriter does parse the pattern, but it only should do it once if used correctly I think.

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -969,7 +982,7 @@ graph(%self, %scale, %zero_point, %dtype):
     return (%weight_quant) )";
   SubgraphRewriter rewriter;
-  rewriter.RegisterRewritePattern(pattern, replacement);
+  rewriter.RegisterRewritePattern(pattern.pattern_string, replacement);
```
Here we would actually reparse the string in subgraph-rewriter, we probably should not be changing this place. And if we want to cache this parsing, we should just keep the rewriter object alive.

jamesr66a replied:
The redundant parsing is an issue whether we cache the results or not. I'm going to go ahead and cache the SubgraphRewriter. That way, there's still a redundant parse, but all of the work is memoized

ZolotukhinM replied:
What I'm saying is that this particular change only makes it worse: we now parse it when for memoization **and** in subgraph-rewriter.

Also I think using subgraph rewriter here is suboptimal anyway and we should just use the matcher directly - in fact we're already doing it with the same pattern just 10 lines earlier. The only thing it is used for here is to remove one node from the matched subgraph - just do it directly in the loop above.

jamesr66a replied:
@ZolotukhinM before this change, we had one `parseIR` call at line 936, and one `parseIR` call embedded in the `runOnGraph` call at line 973 (by way of `runOnGraph -> rewriteSinglePatternOnGraph ->parseIR`). These two callsites were hit every time `FoldQuantizeCallIntoBuffer` was called. My change makes it so that the first such call is done once and stored away, leaving only the second call within `runOnGraph`.

I agree with your second point, I can just switch this over and be done with it

ZolotukhinM replied:
Right, makes sense. I misunderstood your earlier point and thought you were claiming savings in subgraph-rewriter itself.

Commit:
Update on "[quantization] Memoize parseIR calls in graph mode quantization"


We don't need to re-parse the IR on every call for these sites.

followup: can we extend this to places we use SubgraphMatcher?

Differential Revision: [D18625743](https://our.internmc.facebook.com/intern/diff/D18625743)

[ghstack-poisoned]

Commit:
Update on "[quantization] Memoize parseIR calls in graph mode quantization"


We don't need to re-parse the IR on every call for these sites.

followup: can we extend this to places we use SubgraphMatcher?

Differential Revision: [D18625743](https://our.internmc.facebook.com/intern/diff/D18625743)

[ghstack-poisoned]

Commit:
Update on "[quantization] Memoize parseIR calls in graph mode quantization"


We don't need to re-parse the IR on every call for these sites.

followup: can we extend this to places we use SubgraphMatcher?

Differential Revision: [D18625743](https://our.internmc.facebook.com/intern/diff/D18625743)

[ghstack-poisoned]

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -151,6 +171,34 @@ class InsertObserversHelper {
   // every time to find a unique id.
   int uid_ = 0;
+
+  // These are the IR patterns that we match when we insert observers.
```
these are the patterns we match to skip inserting observers

jamesr66a replied:
oh good point, thanks

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -1009,177 +1030,190 @@ void InsertPrepackUnpack(script::Module& module) {
-  GRAPH_DUMP(
-      "Before FoldPrepackWeightIntoModule: ",
-      graph);
+struct FoldPrepackedWeightIntoModuleHelper {
```
there seems to be more changes which didn't match the description of the commit anymore, please split the changes next time

Commit:
Update on "[quantization] Memoize parseIR calls in graph mode quantization"


We don't need to re-parse the IR on every call for these sites.

followup: can we extend this to places we use SubgraphMatcher?

Differential Revision: [D18625743](https://our.internmc.facebook.com/intern/diff/D18625743)

[ghstack-poisoned]

Commit:
Update on "[quantization] Memoize parseIR calls in graph mode quantization"


We don't need to re-parse the IR on every call for these sites.

followup: can we extend this to places we use SubgraphMatcher?

Differential Revision: [D18625743](https://our.internmc.facebook.com/intern/diff/D18625743)

[ghstack-poisoned]
